### PR TITLE
[api-minor] Add a jpx decoder based on OpenJPEG 2.5.2

### LINKED ISSUE #17289: PDF.js doesn't render large images in PDF that other renderers do
Link to PDF file: [example.pdf](https://github.com/mozilla/pdf.js/files/13391396/example.pdf)

- Firefox 119.0.1 (64-bit), Chromium 119.0.6045.159 (Official Build) snap (64-bit)
- Ubuntu Linux 23.04.
- PDF.js version: 4.0.189 [50f52b43a]
- Is a browser extension: no.

Steps to reproduce the problem:
1. Just open the PDF on PDF.js

What is the expected behavior?
Notice the scanned book page image gets rendered:
image
Screenshot take from Atril Document Viewer.

What went wrong?
Notice the scanned book page does not get rendered here:
image

Console log:
PDF 247b3d402477644b1b7b61c8c1311da0 [1.6 PDF24 / PDF24] (PDF.js: 4.0.189 [50f52b43a]) [viewer.mjs:12052:13]
Warning: Unsupported header type 1970628964 (uuid). 2 [pdf.worker.mjs:339:13]
Warning: JPX: Unsupported COD options (terminationOnEachCodingPass, verticallyStripe, predictableTermination). 08:55:09.017 Warning: Unable to decode image "img_p0_1": "JpxError: JPX error: JPX error: Invalid tag tree".
Warning: Dependent image isn't ready yet

Last three lines repeat for every scanned page image.

### PR DESCRIPTION
**Pull request by calixteman**
The decoder is compiled in WASM:
https://github.com/mozilla/pdf.js.openjpeg

Fixes #17289, #17061, #16485, #13051, #6365, #4648, #12213.

### CODE REVIEW

File: src/core/openjpeg_decoder.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    this.#module = module;
+  }
+
+  static decode(data) {
+    this.#instance ||= new OpenJPEGDecoder(OpenJPEG());
+    return this.#instance.#decode(data);
+  }
+
+  #decode(data) {
+    const mod = this.#module;
```
Comment (timvandermeij) [inline src/core/openjpeg_decoder.js:40]: Let's call this `const module = ...` to avoid confusion with e.g. modulo.

File: src/core/openjpeg_decoder.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        properties.height = Ysiz - YOsiz;
+        properties.componentsCount = Csiz;
+        return properties;
+      }
+    }
+    throw new JpxError("No size marker found in JPX stream");
+  }
+}
+
+export { JpxError, OpenJPEGDecoder };
```
Comment (timvandermeij) [inline src/core/openjpeg_decoder.js:102]: The `JpxError` exception doesn't seem to actually be caught anywhere since before it was also only used in `jpx.js` which is now removed. While it makes sense to raise a specific exception here internally in this file, I don't think we need to export it here?

Note that the fuzzers only use the exception message, see https://github.com/mozilla/pdf.js/blob/master/test/fuzz/jpx_image.fuzz.js#L10-L14, so they also don't need the exception class itself, and the image decoders bundle also didn't expose them. However, depending on what we decide to do with the JPX fuzzer in the other comment, we might need to change the fuzzer implementation at https://github.com/mozilla/pdf.js/blob/master/test/fuzz/jpx_image.fuzz.js#L10 because I can imagine that the "Cannot read properties" message might not be emitted by OpenJPEG anymore and something else is logged instead.

File: src/core/openjpeg_decoder.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import { BaseException } from "../shared/util.js";
+import OpenJPEG from "../../external/openjpeg/openjpeg.js";
+
+class JpxError extends BaseException {
+  constructor(msg) {
+    super(`JPX error: ${msg}`, "JpxError");
+  }
+}
+
+class OpenJPEGDecoder {
```
Comment (timvandermeij) [inline src/core/openjpeg_decoder.js:25]: Now that we rely on the upstream OpenJPEG implementation, we might be able to remove the JPX fuzzer (see https://github.com/mozilla/pdf.js/blob/master/test/fuzz/jpx_image.fuzz.js and https://github.com/google/oss-fuzz/blob/master/projects/pdf-js/build.sh#L25 because we no longer use our own decoder. It looks like OpenJPEG is already fuzzed (see https://github.com/uclouvain/openjpeg/tree/master/tests/fuzzers and https://github.com/google/oss-fuzz/tree/master/projects/openjpeg) but it's not entirely clear to me if JPX is part of that (maybe @manunio can chime in here given that they introduced our fuzzers in #17337 and might know more about this).

File: src/pdf.image_decoders.js [hunk 4]
```diff
@@ -20,7 +20,7 @@ import {
 } from "./shared/util.js";
 import { Jbig2Image } from "./core/jbig2.js";
 import { JpegImage } from "./core/jpg.js";
-import { JpxImage } from "./core/jpx.js";
```
Comment (timvandermeij) [inline src/pdf.image_decoders.js:23]: Removing `JpxImage` from the exports makes this a breaking change because `JpxImage` is exposed publicly. Note how for example the fuzzer relies on it at https://github.com/mozilla/pdf.js/blob/master/test/fuzz/jpx_image.fuzz.js#L2 but also the API test at https://github.com/mozilla/pdf.js/blob/7bd7c6ca5a68e1f35bfd586cf2e4c14890ac920d/test/unit/pdf.image_decoders_spec.js#L23. If we do this, we might need to bump the major version (but if we do it might make sense to also replace the other custom image decoders with OpenJPEG first where possible to do it all in one go; see the other comment about that below).

File: test/unit/pdf.image_decoders_spec.js [hunk 5]
```diff
@@ -37,7 +37,7 @@ describe("pdfimage_api", function () {
       getVerbosityLevel,
       Jbig2Image,
       JpegImage,
-      JpxImage,
```
Comment (timvandermeij) [inline test/unit/pdf.image_decoders_spec.js:40]: Could OpenJPEG also be used to e.g. replace our custom JPEG decoder in `jpg.js` and therefore the `JpegImage` class as well? Obviously that'd not be in scope of this PR, but I wanted to ask because that'd be really nice and also be a factor for determining when we'll do the major bump from the comment above. (I don't think OpenJPEG does JBIG2 but maybe we can also replace that one with a better upstream version later.)

**Review by timvandermeij — CHANGES_REQUESTED** [review]
I'm positive about replacing our own decoder with OpenJPEG: it fixes a whole lot of issues, it reduces maintenance and the upstream project seems well-tested, widely used and more performing because it's written in C. I have looked at most of the code, but not everything in full detail yet because the code might change when these first comments are addressed. Overall this is really nice work so far!

File: src/core/openjpeg_decoder.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      data_ptr,
+      data.length,
+      image_ptr_ptr,
+      image_size_ptr
+    );
+    mod._free(data_ptr);
+    if (ret) {
+      mod._free(image_ptr_ptr);
+      mod._free(image_size_ptr);
+      return new Uint8Array(0);
```
Comment (Snuffleupagus) [inline src/core/openjpeg_decoder.js:56]: This means that image decoding failed, right?
If so, we should throw a `JpxError` rather than returning dummy-data!

File: src/core/openjpeg_decoder.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    const image_ptr = mod.HEAPU32[image_ptr_ptr / 4];
+    const image_size = mod.HEAPU32[image_size_ptr / 4];
+    const newArray = mod.HEAPU8.slice(image_ptr, image_ptr + image_size);
+
+    mod._free(image_ptr);
+    mod._free(image_ptr_ptr);
+    mod._free(image_size_ptr);
+
+    return newArray;
```
Comment (Snuffleupagus) [inline src/core/openjpeg_decoder.js:67]: What sort of TypedArray is this?
To maintain consistency with the other image decoders, and the comment just below, this should be wrapped in an `Uint8ClampedArray` (unless it's already that format).

File: test/unit/pdf.image_decoders_spec.js (hunk 5 above)
Comment (Snuffleupagus) [reply test/unit/pdf.image_decoders_spec.js:40]: I *think* that this decoder is only for JPEG2000, given https://github.com/uclouvain/openjpeg?tab=readme-ov-file#what-is-openjpeg-

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: src/core/openjpeg_decoder.js (hunk 3 above)
Comment (timvandermeij) [reply src/core/openjpeg_decoder.js:25]: Coming to think of this again, the https://github.com/mozilla/pdf.js.openjpeg repository does contain some wrapping code for OpenJPEG which I guess could be useful to fuzz. I'm not sure how useful that is given that OpenJPEG itself is already fuzzed upstream and we're only wrapping it thinly, but if it is we might not actually have to remove the fuzzer because it's then still fuzzing some of our code (perhaps we only need to update the exception message from the other comment).

File: src/core/openjpeg_decoder.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  #module = null;
+
+  static #instance = null;
+
+  constructor(module) {
+    this.#module = module;
+  }
+
+  static decode(data) {
+    this.#instance ||= new OpenJPEGDecoder(OpenJPEG());
```
[the remark this answers is not in the document]
Comment (calixteman) [reply src/core/openjpeg_decoder.js]: As a side note: right now the wasm data is inlined as base64 string in the generated `openjpeg.js`.
This file is ~317Kb but in setting `SINGLE_FILE=0` (the build works well on Windows with Docker installed):
https://github.com/mozilla/pdf.js.openjpeg/blob/main/compile.sh#L23
it's possible to get two files: a js ~10Kb and a wasm ~230Kb.
It works well locally but when trying to build for m-c, webpack adds a `document.baseUri` in the worker code which makes it failing and I didn't manage to get rid of it.

File: src/core/openjpeg_decoder.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    return this.#instance.#decode(data);
+  }
+
+  static cleanup() {
+    this.#instance = null;
+  }
+
+  #decode(data) {
+    const module = this.#module;
+    const data_ptr = module._malloc(data.length);
```
Comment (timvandermeij) [inline src/core/openjpeg_decoder.js:45]: I'm not sure how feasible this is, but figured I'd ask just in case: would it be possible to provide a wrapper function in e.g. https://github.com/mozilla/pdf.js.openjpeg/blob/main/src/decoder.c that does most of this function's logic, but then in C? Looking at the code in this method it feels a bit like C details like explicit malloc/free are leaking through to JavaScript, and I'm not sure if that's wanted or not. If we could expose a `jpx_decode(data)` API from `decoder.c` we might be able to move all those details into the C code, which would simplify the JavaScript code here and keep e.g. malloc/free concerns in C where they are relevant.

**Comment by calixteman** [comment]
I've a fix for the test failure, I'll amend tomorrow.

**Comment by Snuffleupagus** [comment]
Given that you've moved more functionality to the "external" code, can't we simplify and reduce the size of the `JpxImage` class even further now? (Attaching a diff, since trying to comment inline didn't work well in this case.)

```diff
diff --git a/src/core/jpx.js b/src/core/jpx.js
index b4494bbd4..af24fe596 100644
--- a/src/core/jpx.js
+++ b/src/core/jpx.js
@@ -23,24 +23,11 @@ class JpxError extends BaseException {
 }
 
 class JpxImage {
-  #module = null;
-
-  static #instance = null;
-
-  constructor(module) {
-    this.#module = module;
-  }
+  static #module = null;
 
   static decode(data, ignoreColorSpace) {
-    this.#instance ||= new JpxImage(OpenJPEG());
-    return this.#instance.#decode(data, ignoreColorSpace);
-  }
-
-  static cleanup() {
-    this.#instance = null;
-  }
+    this.#module ||= OpenJPEG();
 
-  #decode(data, ignoreColorSpace) {
     const imageData = this.#module.decode(data, ignoreColorSpace);
     if (!imageData) {
       throw new JpxError("JPX decode failed");
@@ -48,6 +35,10 @@ class JpxImage {
     return imageData;
   }
 
+  static cleanup() {
+    this.#module = null;
+  }
+
   static parseImageProperties(stream) {
     // No need to use OpenJPEG here since we're only getting very basic
     // information which are located in the first bytes of the file.
```

File: src/core/image.js [hunk 10]
```diff
@@ -205,6 +207,10 @@ class PDFImage {
         localColorSpaceCache,
       });
       this.numComps = this.colorSpace.numComps;
+
+      // If the jpx image has a color space then it musn't be used in order to
+      // be able to use the color space that comes from the pdf.
+      this.ignoreColorSpace = jpxDecode && this.colorSpace?.name === "Indexed";
```
Comment (Snuffleupagus) [inline src/core/image.js:213]: Nit: Can't we skip the optional chaining *this case*, since `this.colorSpace` is explicitly set just above?
```suggestion
      this.ignoreColorSpace = jpxDecode && this.colorSpace.name === "Indexed";
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, since the remaining test movement seem more-or-less inline with what you'd "expect" when changing image decoder; thank you!

While this increases the bundle size slightly, we do get at least two benefits:
 - Fixing a whole bunch of old image conversion issues all at once.
 - Slightly reduces the amount of code that we need to maintain ourselves.

(Given all of the JPEG 2000 features that *our* old decoder didn't support, if we'd implement a complete one ourselves the code-size would probably have become a fair bit larger anyway.)

File: src/core/jpx.js [hunk 11]
```diff
@@ -137,2213 +62,14 @@ class JpxImage {
...
-        this.width = Xsiz - XOsiz;
-        this.height = Ysiz - YOsiz;
-        this.componentsCount = Csiz;
-        // Results are always returned as `Uint8ClampedArray`s.
-        this.bitsPerComponent = 8;
-        return;
+        properties.width = Xsiz - XOsiz;
+        properties.height = Ysiz - YOsiz;
+        properties.componentsCount = Csiz;
+        return properties;
```
Comment (Snuffleupagus) [inline src/core/jpx.js:68]: Just one more question, given the immediate return here: Can't we simply inline the object instead?
```suggestion
        return {
          width: Xsiz - XOsiz,
          height: Ysiz - YOsiz,
          // Results are always returned as `Uint8ClampedArray`s.
          bitsPerComponent: 8,
          componentsCount: Csiz,
        };
```

**Comment by timvandermeij** [comment]
I also think this version of the patch is really nice; thank you for doing this!

I only have one question, which may or may not need a follow-up issue/PR depending on which guarantees we give about the image decoder package's API stability. Previously, from what I can gather, for all image decoders we basically guaranteed one public method was always available and served as the only entrypoint to the module, namely `parse(data)`. In this PR we no longer do that for `jpx.js` only since that has become `decode(data)`. However, it could be argued that the previous APIs weren't really defined properly because for example `jpg.js` exposed more public methods that look like they were intended to be private instead, and other decoders didn't expose those as well. On the other hand, we could probably limit breakage even more by renaming `decode(data)` to `parse(data)` so the API is as similar as possible as before, to reduce the risk of issues being opened for this or it being regarded as an API-breaking change.

@calixteman @Snuffleupagus What do you think? Would it be a good idea to rename `decode(data)` to `parse(data)` for better API compatibility of the image decoders package, or is this something we never really defined a formal public API for?

**Comment by timvandermeij** [comment]
Coming to think of this again, I actually think we should change this, not only to make the API equal to before but also because e.g. our JPX fuzzer relies on this API; see https://github.com/mozilla/pdf.js/blob/master/test/fuzz/jpx_image.fuzz.js#L21. I'll try to prepare a patch for that today.

**Comment by calixteman** [comment]
I tend to overlook the non-Firefox use of pdf.js, sorry about that.
That said I'm fine with whatever name change.
About the fuzzing stuff: we didn't have any news from the people who added it... and I have no idea if it's because they didn't find anything or if it's because kind of dead...

**Comment by Snuffleupagus** [comment]
> I think we could quite easily avoid any breakage of the "official" `parse(data)` API by renaming `decode(data)` to `parse(data)`, to reduce the risk of issues being opened for this or it being regarded as an API-breaking change.

A simple renaming would unfortunately not be sufficient, since all the methods are now `static`, and we'd need to make additional changes in the code as well.

**Comment by timvandermeij** [comment]
Yes, that is indeed unfortunate. I have managed to write a test to assert that the expected public API is present and that now correctly fails on `JpxImage`, but now I'm wondering what we should do here because we can't get the API _fully_ equal to before and the current version with static methods arguably also makes more sense in general.

For example, previously `parse(data)` would expose the parsed data as public members on the `JpxImage` object, but now `parse(data)` directly returns the parsed data (which in my opinion is also better/easier). Moreover, making it a static method avoids needing an object, which previously also only really seemed necessary because of the storing of parsed data on it.

Even if we were to revert to an object without static methods, it still wouldn't be a 100% backwards-compatible change because we don't have e.g. the `tiles` member from https://github.com/mozilla/pdf.js/blob/fb0775525e62c0bc46bc2eea03f1df129b5ed1f9/src/core/jpx_stream.js#L54 anymore.

I'm not sure what's best here. I see a couple of options:

- Decide that the current API is better and promote this change to `[api-major]`. In this case I do think we need to update the other decoders similarly, so with static methods, as well so that they are all in sync again (because having different APIs will be confusing). I'll also provide the extended unit test then so that from that point on the public API is documented and can't accidentally regress anymore.
- Revert to non-static methods for `JpxImage` and re-expose the parsed data public members (and accept that e.g. `tiles` will be gone, but the chance of people actually using it are probably small anyway, but technically this still breaks the API).

I don't really see a point in doing the second option in combination with letting `parse(data)` return the data because then there is really no need to construct an object and the absence of the members would be equally API-breaking.

In short, all options are API-breaking, but some to a lesser extent. I'm wondering if for all of them `[api-minor]` is a good label here, and if we go to `[api-major]` we might as well bite the bullet and align the other decoders too while we're at it.

What do you think? I'm tempted to go for option one here, but it will be a bit more work.

**Comment by timvandermeij** [comment]
That's a fair point. I also don't think there are many users, and besides I'm not sure if we ever really made any API guarantees for this code anyway. I'm probably overthinking the problem here, and we can always make further adjustments later if the need arises.

In that case I'll only make a patch to fix the fuzzer code because that is currently broken after this PR, and after that all code in this repository should work again. I do think the new `JpxImage` API is better in general, and even the `decode` name is probably an improvement over the previous `parse` name because we in fact _decode_ the full image data but we only _parse_ some basic image properties from the data blob (so https://github.com/mozilla/pdf.js/blob/master/src/core/jpx.js#L41 is named correctly in my opinion).

Thanks for the feedback!
